File: contests/lc/756.PyramidTransitionMatrix/solution.go
package main

func pyramidTransition(bottom string, allowed []string) bool {
	pairs := make(map[string][]byte)
	for _, rule := range allowed {
		pair := rule[:2]
		pairs[pair] = append(pairs[pair], rule[2])
	}

	memo := make(map[string]bool)

	var canBuild func(row string) bool
	canBuild = func(row string) bool {
		if len(row) == 1 {
			return true
		}
		if val, ok := memo[row]; ok {
			return val
		}

		next := make([]byte, len(row)-1)
		var buildNext func(pos int) bool
		buildNext = func(pos int) bool {
			if pos == len(row)-1 {
				return canBuild(string(next))
			}

			pair := row[pos : pos+2]
			choices, ok := pairs[pair]
			if !ok {
				return false
			}
			for _, ch := range choices {
				next[pos] = ch
				if buildNext(pos + 1) {
					return true
				}
			}
			return false
		}

		res := buildNext(0)
		memo[row] = res
		return res
	}

	return canBuild(bottom)
}

File: contests/lc/756.PyramidTransitionMatrix/solution.py
from collections import defaultdict
from typing import List


class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        pairs = defaultdict(list)
        for rule in allowed:
            pairs[rule[:2]].append(rule[2])

        memo = {}

        def can_build(row: str) -> bool:
            if len(row) == 1:
                return True
            if row in memo:
                return memo[row]

            next_row = [""] * (len(row) - 1)

            def build_next(pos: int) -> bool:
                if pos == len(row) - 1:
                    return can_build("".join(next_row))

                pair = row[pos : pos + 2]
                if pair not in pairs:
                    return False
                for ch in pairs[pair]:
                    next_row[pos] = ch
                    if build_next(pos + 1):
                        return True
                return False

            memo[row] = build_next(0)
            return memo[row]

        return can_build(bottom)
